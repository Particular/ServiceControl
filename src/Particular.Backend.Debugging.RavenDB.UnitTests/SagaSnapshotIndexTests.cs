using System;
using System.Collections.Generic;
using NUnit.Framework;
using ObjectApproval;
using Particular.Backend.Debugging.Api;
using Particular.Backend.Debugging.RavenDB.Api;
using Particular.Backend.Debugging.RavenDB.Model;

namespace Particular.Backend.Debugging.RavenDB.UnitTests
{
    [TestFixture]
    class SagaSnapshotIndexTests
    {
        [Test]
        public void RunMapReduce()
        {
            using (var store = InMemoryStoreBuilder.GetInMemoryStore())
            {
                store.ExecuteIndex(new SagaSnapshotIndex());
                using (var session = store.OpenSession())
                {
                    foreach (var sagaHistory in GetFakeHistory())
                    {
                        session.Store(sagaHistory);
                    }
                    session.SaveChanges();
                }

                store.WaitForIndexing();

                using (var session = store.OpenSession())
                {
                    SagaHistory history;
                    DateTime lastModified;
                    SagaSnapshotIndex.TryGetSagaHistory(session, Guid.Empty, out history, out lastModified);
                    Assert.AreEqual(new DateTime(2002, 4, 1, 1, 1, 1, DateTimeKind.Utc),lastModified);
                    ObjectApprover.VerifyWithJson(history);
                }
            }
        }

        static IEnumerable<object> GetFakeHistory()
        {
            yield return new SagaSnapshot
            {
                SagaId = Guid.Empty,
                SagaType = "MySaga1",
                Endpoint = "MyEndpoint",
                FinishTime = new DateTime(2001, 2, 1, 1, 1, 1, DateTimeKind.Utc),
                Status = SagaStateChangeStatus.Updated,
                StartTime = new DateTime(2001, 1, 1, 1, 1, 1, DateTimeKind.Utc),
                StateAfterChange = "Updated"
            };
            yield return new SagaSnapshot
            {
                SagaId = Guid.Empty,
                SagaType = "MySaga1",
                Endpoint = "MyEndpoint",
                FinishTime = new DateTime(2002, 4, 1, 1, 1, 1, DateTimeKind.Utc),
                Status = SagaStateChangeStatus.Completed,
                StartTime = new DateTime(2002, 3, 1, 1, 1, 1, DateTimeKind.Utc),
                StateAfterChange = "Completed"
            };
        }

    }
}