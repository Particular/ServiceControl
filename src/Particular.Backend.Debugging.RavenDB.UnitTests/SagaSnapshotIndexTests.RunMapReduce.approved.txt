{
    "Id": "00000000-0000-0000-0000-000000000000",
    "SagaId": "00000000-0000-0000-0000-000000000000",
    "SagaType": "MySaga1",
    "Changes": [
        {
            "StartTime": "2002-03-01T01:01:01Z",
            "FinishTime": "2002-04-01T01:01:01Z",
            "Status": "Completed",
            "StateAfterChange": "Completed",
            "InitiatingMessage": null,
            "OutgoingMessages": [],
            "Endpoint": "MyEndpoint"
        },
        {
            "StartTime": "2001-01-01T01:01:01Z",
            "FinishTime": "2001-02-01T01:01:01Z",
            "Status": "Updated",
            "StateAfterChange": "Updated",
            "InitiatingMessage": null,
            "OutgoingMessages": [],
            "Endpoint": "MyEndpoint"
        }
    ]
}